### PR TITLE
Proposal: Add method to remove a queue from recovery without deleting it

### PR DESCRIPTION
**Pull request by vikinghawk**
## Proposed Changes
This is a proposal to add a new method to address the following usecase:
1. Connection A declares a durable, shared,non-autodelete queue "ABC"
2. Connection A consumes from the queue
3. Connection B re-declares the same queue "ABC" as it does not know or care whether it already exists from Connection A
4. Connection B consume from the queue
5. Connection A cancels its queue consumers
6. Connection A does not want to delete the queue because it may still be in use by others, but it would like to remove the queue from its list of recovery items because it is no longer actively using it.

This adds a public method to AutorecoveringConnection to allow for removal of a queue from recorded queues (and its related bindings) without having to call Channel.queueDelete().

Let me know if this is something you guys would be willing to pull in. I was originally looking at using reflections to get to the necessary methods/variables to do this. But obviously that is fragile for me to maintain and there may be others in the community with a similar usecase that could benefit.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [ ] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc.

### CODE REVIEW

File: src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java [hunk 1]
```diff
@@ -761,6 +761,28 @@ void deleteRecordedQueue(String queue) {
...
     }
+    
+    /**
+     * Remove the queue from list of recorded queues to recover after connection failure.
+     * Intended to be used in usecases where you want to remove the queue from this connection's recovery list but don't want to delete the queue from the server.
+     * 
+     * @param queue queue name to remove from recorded recovery queues
+     * @param ifUnused if true, the RecordedQueue will only be removed if no more consumers from this connection are using it.
+     */
+    public void removeRecordedQueue(final String queue, final boolean ifUnused) {
```
Comment (michaelklishin) [inline src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java]: WDYT about `excludeQueueFromRecovery` or `doNotRecoveryQueue`?

**Comment by michaelklishin** [comment]
The scenario sounds reasonable to me. I'm not sure the name is what we should go with (suggested a couple of alternatives above) but otherwise no objections from me @acogoluegnes.

File: src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java (hunk 1 above)
Comment (vikinghawk) [reply src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java]: I like doNotRecoverQueue or removeQueueFromRecovery.
excludeQueueFromRecovery works too. I think we would just need to make it clear you have to call the method after declaring a queue and not before. Its not adding an exclude filter or anything like that for future declareQueue calls.

Comment (michaelklishin) [reply src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java]: My personal favorite is `excludeQueueFromRecovery`.

Comment (acogoluegnes) [reply src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java]: I'd go with `excludeQueueFromRecovery`.

Comment (vikinghawk) [reply src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java]: works for me :)
I'll change that and look at adding a junit.
Thanks guys.

**Comment by vikinghawk** [comment]
Changes made and moved to 4.x.x branch on #346

**Pull request closed without merging**